contracts/bonding/BondingCheckpoints.sol

```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/governance/Governor.sol";
+import "@openzeppelin/contracts/governance/extensions/GovernorSettings.sol";
+import "@openzeppelin/contracts/utils/Checkpoints.sol";
+
+import "./libraries/EarningsPool.sol";
+import "./libraries/EarningsPoolLIP36.sol";
+
+import "../IController.sol";
+import "../rounds/IRoundsManager.sol";
+import "./BondingManager.sol";
+
+contract BondingCheckpoints {
+    using Checkpoints for Checkpoints.Trace224;
+
+    uint256 public constant MAX_ROUNDS_WITHOUT_CHECKPOINT = 100;
+
+    BondingManager public immutable bondingManagerAddr;
+
+    constructor(BondingManager _bondingManager) {
+        bondingManagerAddr = _bondingManager;
+    }
+
+    // We can't lookup the "checkpoint time" from the Checkpoints lib, only the
+    // current value. So instead of checkpointing the bonded amount and
+    // delegatee we snapshot the start round of each delegator change and lookup
+    // the specific values on the separate delegatorSnapshots mapping.
+    // TODO: Consider writing our own checkpoints lib version instead that
+    // stores directly the data we want inline.
```

Comment:
I've decided to do this as the curr implementation is a bit weird. Should be simple enough to implement using a separate "snapshotted rounds" array and the same mapping of round to `DelegatorInfo`. Instead of looking up the round on the hacked `Checkpoints.Trace224`, we lookup the round with [`Arrays.findUpperBound()`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Arrays.sol#L24) and then access the map.

Reply:
Done here: https://github.com/livepeer/protocol/pull/611/commits/1cbe31d866e5bc8c6ba29cb97f05446093b33ffa